File: srclab-common/src/main/java/xyz/srclab/common/base/Provider.java
package xyz.srclab.common.base;

import xyz.srclab.annotation.Nullable;
import xyz.srclab.common.reflect.ClassKit;

/**
 * @author sunqian
 */
public interface Provider<T> {

    static <T> Provider<T> of(Class<?> providerClass) {
        if (!Provider.class.isAssignableFrom(providerClass)) {
            throw new IllegalArgumentException("Class is not a Provider: " + providerClass);
        }
        return ClassKit.newInstance(providerClass);
    }

    static <T> Provider<T> of(String className) {
        return of(className, Loader.currentClassLoader());
    }

    static <T> Provider<T> of(String className, ClassLoader classLoader) {
        @Nullable Class<?> providerClass = Loader.findClass(className, classLoader);
        if (providerClass == null) {
            throw new IllegalArgumentException("Class not found: " + className);
        }
        return of(providerClass);
    }

    static <T> T get(String candidateClasses) {
        String[] classNames = candidateClasses.split(",");
        for (String className : classNames) {
            Provider<T> provider = of(className);
            @Nullable T product = provider.provide();
            if (product != null) {
                return product;
            }
        }
        throw new IllegalArgumentException("No available provider: " + candidateClasses);
    }

    @Nullable
    T provide();
}

File: boat-common/src/main/kotlin/xyz/srclab/common/base/Provider.kt
package xyz.srclab.common.base

import xyz.srclab.common.reflect.findClassToInstance

interface Provider<S, T : Any> {

    fun parse(spec: S): List<T>

    @JvmDefault
    fun parseFirst(spec: S): T {
        return parseFirstOrNull(spec) ?: throw IllegalArgumentException("Illegal provider specification: $spec")
    }

    fun parseFirstOrNull(spec: S): T?

    companion object {

        @JvmStatic
        @JvmOverloads
        fun <T : Any> charsProvider(strict: Boolean = false): CharsProvider<T> {
            return if (strict) StrictCharsProvider.ofType() else CharsProvider.ofType()
        }

        @JvmStatic
        @JvmOverloads
        @JvmName("parseChars")
        fun <T : Any> CharSequence.parseCharsProviders(strictly: Boolean = false): List<T> {
            return charsProvider<T>(strictly).parse(this)
        }

        @JvmStatic
        @JvmOverloads
        @JvmName("parseCharsFirst")
        fun <T : Any> CharSequence.parseCharsFirstProvider(strictly: Boolean = false): T {
            return charsProvider<T>(strictly).parseFirst(this)
        }

        @JvmStatic
        @JvmOverloads
        @JvmName("parseCharsFirstOrNull")
        fun <T : Any> CharSequence.parseCharsFirstProviderOrNull(strictly: Boolean = false): T? {
            return charsProvider<T>(strictly).parseFirstOrNull(this)
        }
    }
}

open class CharsProvider<T : Any> : Provider<CharSequence, T> {

    override fun parse(spec: CharSequence): List<T> {
        val classNames = spec.split(",")
        val result = ArrayList<T>(classNames.size)
        for (className in classNames) {
            val trimmedClassName = className.trim()
            val product: T? = try {
                trimmedClassName.findClassToInstance()
            } catch (e: Exception) {
                continue
            }
            if (product !== null) {
                result.add(product)
            }
        }
        return result
    }

    override fun parseFirstOrNull(spec: CharSequence): T? {
        val classNames = spec.split(",")
        for (className in classNames) {
            val trimmedClassName = className.trim()
            val product: T? = try {
                trimmedClassName.findClassToInstance()
            } catch (e: Exception) {
                continue
            }
            if (product !== null) {
                return product
            }
        }
        return null
    }

    companion object {

        private val INSTANCE = CharsProvider<Any>()

        @JvmStatic
        fun <T : Any> ofType(): CharsProvider<T> {
            return INSTANCE.asAny()
        }
    }
}

class StrictCharsProvider<T : Any> : CharsProvider<T>() {

    override fun parse(spec: CharSequence): List<T> {
        val classNames = spec.split(",")
        val result = ArrayList<T>(classNames.size)
        for (className in classNames) {
            result.add(createInstance(className))
        }
        return result
    }

    override fun parseFirstOrNull(spec: CharSequence): T? {
        val classNames = spec.split(",")
        for (className in classNames) {
            return createInstance(className)
        }
        return null
    }

    private fun createInstance(className: CharSequence): T {
        val trimmedClassName = className.trim()
        val product: T? = try {
            trimmedClassName.findClassToInstance()
        } catch (e: Exception) {
            throw IllegalStateException("Instantiate class $trimmedClassName failed.", e)
        }
        if (product === null) {
            throw IllegalStateException("Class $trimmedClassName was not found.")
        }
        return product
    }

    companion object {

        private val INSTANCE = StrictCharsProvider<Any>()

        @JvmStatic
        fun <T : Any> ofType(): StrictCharsProvider<T> {
            return INSTANCE.asAny()
        }
    }
}